rpcs3/Emu/RSX/Common/texture_cache_predictor.h

```diff
@@ -0,0 +1,389 @@
+
+		u32 confidence;
+
+	public:
+		texture_cache_predictor_entry()
+		{
+			reset();
+		};
+		~texture_cache_predictor_entry() = default;
+
+		void update_confidence(s32 delta)
+		{
+			if (delta > 0)
+			{
+				confidence += delta;
+
+				if (confidence > max_confidence)
+				{
+					confidence = max_confidence;
+				}
+			}
+			else if (delta < 0)
+			{
+				u32 neg_delta = static_cast<u32>(-delta);
+				if (confidence > neg_delta)
+				{
+					confidence -= neg_delta;
+				}
+				else
+				{
+					confidence = 0;
+				}
+			}
+		}
+
+		u32 get_confidence() const
+		{
+			return confidence;
+		}
+
+	private:
+		// Returns how many writes we think there will be this time (i.e. between the last flush and the next flush)
+		// Returning UINT32_MAX means no guess is possible
+		u32 guess_number_of_writes() const
+		{
+			const auto history_size = write_history.size();
+
+			if (history_size == 0)
+			{
+				// We need some history to be able to take a guess
+				return UINT32_MAX;
+			}
+			else if (history_size == 1)
+			{
+				// If we have one history entry, we assume it will repeat
+				return write_history[0];
+			}
+			else
+			{
+				// For more than one entry, we try and find a pattern, and assume it holds
+
+				const u32 stop_when_found_matches = 5;
+				u32 matches_found = 0;
+				u32 guess = UINT32_MAX;
+
+				for (u32 i = 0; i < history_size; i++)
+				{
+					// If we are past the number of writes, it's not the same as this time
+					if (write_history[i] < m_writes_since_last_flush)
+						continue;
+
+					u32 cur_matches_found = 1;
+
+					// Try to find more matches
+					for (u32 j = 0; i + j + 1 < history_size; j++)
+					{
+						if (write_history[i + j + 1] != write_history[j])
+							break;
+
+						// We found another matching value
+						cur_matches_found++;
+
+						if (cur_matches_found >= stop_when_found_matches)
+							break;
+					}
+
+					// If we found more matches than all other comparisons, we take a guess
+					if (cur_matches_found > matches_found)
+					{
+						guess = write_history[i];
+						matches_found = cur_matches_found;
+					}
+
+					if (matches_found >= stop_when_found_matches)
+						break;
+				}
+
+				return guess;
+			}
+		}
+
+		void calculate_next_guess(bool reset)
+		{
+			if (reset || m_guessed_writes == UINT32_MAX || m_writes_since_last_flush > m_guessed_writes)
+			{
+				m_guessed_writes = guess_number_of_writes();
+			}
+		}
+
+	public:
+		void reset()
+		{
+			confidence = starting_confidence;
+			m_writes_since_last_flush = 0;
+			m_guessed_writes = UINT32_MAX;
+			write_history.clear();
+		}
+
+		void on_flush()
+		{
+			update_confidence(is_flush_likely() ? confidence_guessed_flush : confidence_incorrect_guess);
+
+			// Update history
+			write_history.push(m_writes_since_last_flush);
+			m_writes_since_last_flush = 0;
+
+			calculate_next_guess(true);
+		}
+
+		void on_write(bool mispredict)
+		{
+			if(mispredict || is_flush_likely())
```

Comment:
space before `(`
Maybe run clang-format on this file.
```suggestion
			if (mispredict || is_flush_likely())
```